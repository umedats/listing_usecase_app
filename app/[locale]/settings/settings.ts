/**
 * Object which will store the necessary settings required to make SP API calls.
 * Each field in this object corresponds to a field in the settings page.
 *
 * region: AWS Region for creating infrastructure
 * accountId: AWS AccountId for creating infrastructure
 * clientId: Client Id part of the selling partner application LWA credentials
 * clientSecret: Client Secret part of the selling partner application LWA credentials
 * refreshToken: Refresh Token used to get the access token
 * sellingPartnerId: Selling Partner Id
 * sellingPartnerIdType: A selling partner identifier type
 * marketplaceId: The MarketplaceId to be used during the Listing Process
 */
export type Settings = {
  region: string;
  accountId: string;
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  sellingPartnerId: string;
  sellingPartnerIdType: string;
  marketplaceId: string;
};

/**
 * returns true if the settings are incomplete.
 * @param settings
 */
export function areSettingsIncomplete(settings: Settings) {
  return (
    !settings.region ||
    !settings.accountId ||
    !settings.clientId ||
    !settings.clientSecret ||
    !settings.refreshToken ||
    !settings.sellingPartnerId ||
    !settings.sellingPartnerIdType ||
    !settings.marketplaceId
  );
}

export const SELLING_PARTNER_TYPE_KEY_VENDOR_CODE = "Vendor Code";
export const SELLING_PARTNER_TYPE_KEY_MERCHANT_ACCOUNT_ID =
  "Merchant Account Id";
