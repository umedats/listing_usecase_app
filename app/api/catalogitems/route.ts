import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import getSPAPIEndpoint, {
  buildCatalogItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  IDENTIFIER_TYPES_HEADER,
  IDENTIFIERS_HEADER,
  KEYWORDS_HEADER,
  SEARCH_CATALOG_ITEMS_API_DOC_LINK,
  SEARCH_CATALOG_ITEMS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import nextResponse from "@/app/utils/next-response-factory";
import { Item, ItemSearchResults } from "@/app/sdk/catalogItems_2022-04-01";
import { ProductSearchResult } from "@/app/[locale]/create-offer/product-result-type";
import { Settings } from "@/app/[locale]/settings/settings";
import { SPAPIRequestResponse } from "@/app/model/types";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the list of Catalog Items based on identifiers/keywords by calling the SearchCatalogItems SP-API.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getCatalogItemsHandler);
}

async function getCatalogItemsHandler(settings: Settings) {
  const { identifiersType, identifiers, keywords, locale } =
    retrieveRequestParamsFromHeader();
  const optionalParams = constructOptionalParametersForApi(
    settings,
    identifiersType,
    identifiers,
    keywords,
    locale,
  );
  const nextResponseOrSPAPIRequestResponse = await fetchCatalogItems(
    settings,
    optionalParams,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  return fetchItemSearchResults(nextResponseOrSPAPIRequestResponse);
}

function retrieveRequestParamsFromHeader() {
  const requestHeaders = headers();
  const identifiersType = requestHeaders.get(IDENTIFIER_TYPES_HEADER);
  const identifierData = requestHeaders.get(IDENTIFIERS_HEADER);
  const identifiers = identifierData ? identifierData.split(",") : null;
  const keywordsData = requestHeaders.get(KEYWORDS_HEADER);
  const keywords = keywordsData ? keywordsData.split(",") : null;
  const locale = requestHeaders.get(LOCALE_HEADER);

  return { identifiersType, identifiers, keywords, locale };
}

function constructOptionalParametersForApi(
  settings: Settings,
  identifiersType: string | null,
  identifiers: string[] | null,
  keywords: string[] | null,
  locale: string | null,
) {
  const includedData = ["productTypes", "summaries"];

  if (identifiers) {
    const optionalParametersForIdentifiers = {
      identifiers: identifiers,
      identifiersType: identifiersType,
      includedData: includedData,
      locale: locale,
    };

    // Seller ID is mandatory when identifier Type is SKU
    if (identifiersType === "SKU") {
      return {
        sellerId: settings.sellingPartnerId,
        ...optionalParametersForIdentifiers,
      };
    }

    // For non-SKU Identifier Type use case.
    return optionalParametersForIdentifiers;
  }

  if (keywords) {
    return {
      keywords: keywords,
      includedData: includedData,
      locale: locale,
    };
  }
}

async function fetchCatalogItems(settings: Settings, opts: any) {
  // Build Catalog Items API Client.
  const catalogApi = await buildCatalogItemsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: SEARCH_CATALOG_ITEMS_API_NAME,
    apiDocumentationLink: SEARCH_CATALOG_ITEMS_API_DOC_LINK,
    request: {
      marketplaceIds: [settings.marketplaceId],
      ...opts,
    },
    response: {},
  };
  try {
    reqResponse.response = await catalogApi.searchCatalogItems(
      reqResponse.request.marketplaceIds,
      opts,
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

function deriveTitle(item: Item) {
  return item.summaries?.[0]?.itemName;
}

function derivePTD(item: Item) {
  return item.productTypes?.[0]?.productType;
}

function fetchItemSearchResults(reqResponse: SPAPIRequestResponse) {
  const searchResults: ItemSearchResults =
    ItemSearchResults.constructFromObject(reqResponse.response, undefined);

  if (searchResults.numberOfResults && searchResults.numberOfResults > 0) {
    // @ts-ignore
    const items: Item[] = searchResults.items;
    const results: ProductSearchResult[] = items.map((item, index) => {
      const searchResult: ProductSearchResult = {
        title: deriveTitle(item),
        asin: item.asin,
        productType: derivePTD(item),
      };
      return searchResult;
    });
    return nextResponse(
      200,
      "OK",
      serializeToJsonString({
        data: results,
        debugContext: [reqResponse],
      }),
    );
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: [],
      debugContext: [reqResponse],
    }),
  );
}
