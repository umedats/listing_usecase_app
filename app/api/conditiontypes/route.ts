import nextResponse from "@/app/utils/next-response-factory";
import { Settings } from "@/app/[locale]/settings/settings";
import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import {
  ASIN_HEADER,
  GET_LISTING_RESTRICTIONS_API_DOC_LINK,
  GET_LISTING_RESTRICTIONS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildListingsRestrictionsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  Restriction,
  RestrictionList,
} from "@/app/sdk/listingsRestrictions_2021-08-01";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the conditions under which a MCID can list a given ASIN by calling SPAPI.
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getConditionTypesHandler);
}

async function getConditionTypesHandler(settings: Settings) {
  // Validate the various headers expected by the API.
  const validateHeadersResponse = validateHeaderParams();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { asin, locale } = validateHeadersResponse;

  // Retrieve restricted Conditions from SP API
  const nextResponseOrSPAPIResponse = await fetchRestrictions(
    settings,
    asin,
    locale,
  );
  if (nextResponseOrSPAPIResponse instanceof NextResponse) {
    return nextResponseOrSPAPIResponse;
  }
  const restrictedConditionTypes: string[] =
    RestrictionList.constructFromObject(
      nextResponseOrSPAPIResponse.response,
      undefined,
    ).restrictions.map((r: Restriction) => r.conditionType);

  const allowedTypes = Object.values(Restriction.ConditionTypeEnum).filter(
    (c) => !restrictedConditionTypes.includes(c),
  );
  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: allowedTypes,
      debugContext: [nextResponseOrSPAPIResponse],
    }),
  );
}

// Validates the header parameters.
function validateHeaderParams() {
  const requestHeaders = headers();
  const asin = requestHeaders.get(ASIN_HEADER);
  if (!asin) {
    const errorMessage = "The request is missing the asin in the headers";
    console.log(errorMessage);
    return nextResponse(400, errorMessage);
  }

  const locale = requestHeaders.get(LOCALE_HEADER);

  return { asin, locale };
}

/**
 * fetches the condition types from the SP API.
 * @param settings user entered settings
 * @param asin asin
 * @param locale user preferred locale.
 */
async function fetchRestrictions(
  settings: Settings,
  asin: string,
  locale: string | null,
) {
  // Build the Restrictions API Client.
  const restrictionsApi = await buildListingsRestrictionsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: GET_LISTING_RESTRICTIONS_API_NAME,
    apiDocumentationLink: GET_LISTING_RESTRICTIONS_API_DOC_LINK,
    request: {
      asin: asin,
      sellerId: settings.sellingPartnerId,
      marketplaceIds: [settings.marketplaceId],
      conditionType: null,
      reasonLocale: locale ? locale : "",
    },
    response: {},
  };

  try {
    reqResponse.response = await restrictionsApi.getListingsRestrictions(
      reqResponse.request.asin,
      reqResponse.request.sellerId,
      reqResponse.request.marketplaceIds,
      {
        conditionType: reqResponse.request.conditionType,
        reasonLocale: reqResponse.request.reasonLocale,
      },
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}
