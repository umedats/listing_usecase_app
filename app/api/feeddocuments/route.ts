import nextResponse from "@/app/utils/next-response-factory";
import { serializeToJsonString } from "@/app/utils/serialization";
import { NextResponse } from "next/server";
import { Settings } from "@/app/[locale]/settings/settings";
import { headers } from "next/dist/client/components/headers";
import {
  CONTENT_TYPE_HEADER,
  FEED_DOCUMENT_ID_HEADER,
  GET_FEED_DOCUMENT_API_NAME,
  GET_FEED_DOCUMENT_DOC_LINK,
  NO_CACHING_DIRECTIVE,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildFeedsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { FeedDocument } from "@/app/sdk/feeds_2021-06-30";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";
import { unzipSync } from "zlib";
import { SPAPIRequestResponse } from "@/app/model/types";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves a feed document content and returns the content in the API response.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getFeedDocumentContentHandler);
}

async function getFeedDocumentContentHandler(settings: Settings) {
  const requiredParametersResponse =
    await validateAndRetrieveRequiredParametersForGetRequest();

  if (requiredParametersResponse instanceof NextResponse) {
    return requiredParametersResponse;
  }

  const { feedDocumentId } = requiredParametersResponse;

  const nextResponseOrSPAPIRequestResponse = await getFeedDocumentUrl(
    settings,
    feedDocumentId,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  const getFeedDocumentContentResponse = await getFeedDocumentContent(
    FeedDocument.constructFromObject(
      nextResponseOrSPAPIRequestResponse.response,
      undefined,
    ),
    nextResponseOrSPAPIRequestResponse,
  );

  if (getFeedDocumentContentResponse instanceof NextResponse) {
    return getFeedDocumentContentResponse;
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: {
        content: getFeedDocumentContentResponse.feedDocumentContent,
        contentType: getFeedDocumentContentResponse.feedDocumentContentType,
      },
      debugContext: [nextResponseOrSPAPIRequestResponse],
    }),
  );
}

async function validateAndRetrieveRequiredParametersForGetRequest() {
  const validateHeadersResponse =
    validateAndRetrieveRequestParamsFromHeaderForGetRequest();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { feedDocumentId } = validateHeadersResponse;
  return { feedDocumentId };
}

function validateAndRetrieveRequestParamsFromHeaderForGetRequest() {
  const requestHeaders = headers();
  const feedDocumentId = requestHeaders.get(FEED_DOCUMENT_ID_HEADER);
  if (!feedDocumentId) {
    const errorMessage =
      "The request is missing the feedDocumentId in the headers";
    return nextResponse(400, errorMessage);
  }

  return { feedDocumentId };
}

async function getFeedDocumentUrl(settings: Settings, feedDocumentId: string) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const reqResponse = {
    apiName: GET_FEED_DOCUMENT_API_NAME,
    apiDocumentationLink: GET_FEED_DOCUMENT_DOC_LINK,
    request: {
      feedDocumentId: [feedDocumentId],
    },
    response: {},
  };

  try {
    reqResponse.response = await feedsApi.getFeedDocument(feedDocumentId);
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

async function getFeedDocumentContent(
  feedDocument: FeedDocument,
  requestResponse: SPAPIRequestResponse,
) {
  try {
    const fetchDocumentResponse = await fetch(feedDocument.url, {
      cache: NO_CACHING_DIRECTIVE,
    });
    const feedDocumentContentType =
      fetchDocumentResponse.headers.get(CONTENT_TYPE_HEADER);
    let feedDocumentContent = "";
    if (feedDocument.compressionAlgorithm) {
      feedDocumentContent = unzipSync(
        await fetchDocumentResponse.arrayBuffer(),
      ).toString();
    } else {
      feedDocumentContent = await fetchDocumentResponse.text();
    }
    return { feedDocumentContent, feedDocumentContentType };
  } catch (error) {
    // Generally these errors are recoverable.
    const errorMessage = "Error while fetching the feed document content.";
    console.log(errorMessage + " " + error);
    return nextResponse(
      500,
      errorMessage,
      serializeToJsonString({
        debugContext: [requestResponse],
      }),
    );
  }
}
