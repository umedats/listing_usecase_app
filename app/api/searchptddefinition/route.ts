import { NextResponse } from "next/server";
import { Settings } from "@/app/[locale]/settings/settings";
import getSPAPIEndpoint, {
  buildDefinitionsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  ProductTypeList,
  ProductType,
} from "@/app/sdk/definitionsProductTypes_2020-09-01";
import { headers } from "next/dist/client/components/headers";
import {
  SEARCH_DEFINITIONS_PRODUCT_TYPE_API_DOC_LINK,
  SEARCH_DEFINITIONS_PRODUCT_TYPE_API_NAME,
  KEYWORDS_HEADER,
} from "@/app/constants/global";
import nextResponse from "@/app/utils/next-response-factory";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Search product types based on keywords. If keywords is empty, it will return all available product types.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(searchPTDDefinitionHandler);
}

async function searchPTDDefinitionHandler(settings: Settings) {
  const keywords = getKeyWordsFromHeader();

  // Retrieve Search Product Types by Keywords from the SP API.
  const nextResponseOrSPAPIRequestResponse =
    await fetchSearchDefinitionsProductTypes(settings, keywords);

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  const productTypeSearchResult: ProductTypeList =
    ProductTypeList.constructFromObject(
      nextResponseOrSPAPIRequestResponse.response,
      undefined,
    );
  // @ts-ignore
  const productTypes: ProductType[] = productTypeSearchResult.productTypes;
  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: productTypes,
      debugContext: [nextResponseOrSPAPIRequestResponse],
    }),
  );
}

function getKeyWordsFromHeader() {
  const requestHeaders = headers();
  const keywords = requestHeaders.get(KEYWORDS_HEADER);
  if (keywords) {
    return keywords.split(",");
  } else {
    return [];
  }
}

/**
 * fetches the result of search product types from the SP API.
 * @param settings user entered settings
 * @param keywords search keywords
 */
async function fetchSearchDefinitionsProductTypes(
  settings: Settings,
  keywords: string[],
) {
  const definitionsApi = await buildDefinitionsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: SEARCH_DEFINITIONS_PRODUCT_TYPE_API_NAME,
    apiDocumentationLink: SEARCH_DEFINITIONS_PRODUCT_TYPE_API_DOC_LINK,
    request: {
      keywords: keywords,
      marketplaceIds: [settings.marketplaceId],
    },
    response: {},
  };

  try {
    reqResponse.response = await definitionsApi.searchDefinitionsProductTypes(
      reqResponse.request.marketplaceIds,
      {
        keywords: keywords,
      },
    );

    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}
