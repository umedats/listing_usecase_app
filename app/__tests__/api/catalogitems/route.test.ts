import "isomorphic-fetch";
import { GET } from "@/app/api/catalogitems/route";
import { mock, instance, when, reset } from "ts-mockito";
import { getSettings } from "@/app/api/settings/wrapper";
import getSPAPIEndpoint, {
  buildCatalogItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { headers } from "next/dist/client/components/headers";
import { ReadonlyHeaders } from "next/dist/server/web/spec-extension/adapters/headers";
import fetch from "jest-fetch-mock";
import { afterEach } from "@jest/globals";
import {
  IDENTIFIER_TYPES_HEADER,
  IDENTIFIERS_HEADER,
  KEYWORDS_HEADER,
  LOCALE_HEADER,
} from "@/app/constants/global";
import { MOCK_SETTINGS } from "@/app/test-utils/mock-settings";
import {
  mockRejectFetchResponse,
  mockResolveFetchResponse,
} from "@/app/test-utils/mock-fetch";
import { CatalogApi } from "@/app/sdk/catalogItems_2022-04-01";
import {
  MOCK_CATALOG_ITEMS_API_EMPTY_RESPONSE,
  MOCK_CATALOG_ITEMS_API_RESPONSE,
} from "@/app/test-utils/mock-catalog-items-result";
import { verifyStatusAndBody } from "@/app/test-utils/next-response-helper";

const IDENTIFIERS = "B00004SPEC";
const ASIN_IDENTIFIERS_TYPE = "ASIN";
const SKU_IDENTIFIERS_TYPE = "SKU";
const KEYWORDS = "SHIRT";
const US_LOCALE = "en_US";
const MOCK_SPI_API_URL: string = "https://sellingpartnerapi-na.amazon.com";

// @ts-ignore
const mockReadOnlyHeaders: ReadonlyHeaders = mock(ReadonlyHeaders);

const mockedSearchCatalogItems = jest.fn();
const catalogApi: CatalogApi = new CatalogApi(undefined);
catalogApi.searchCatalogItems = mockedSearchCatalogItems;

jest.mock("@/app/sdk/factory/ListingsSPAPIClientsFactory");
const mockedBuildCatalogAPIClient = jest.mocked(buildCatalogItemsAPIClient);
mockedBuildCatalogAPIClient.mockResolvedValue(catalogApi);
const mockedGetSPAPIEndpoint = jest.mocked(getSPAPIEndpoint);

jest.mock("@/app/api/settings/wrapper");
const mockedGetSettings = jest.mocked(getSettings);

jest.mock("next/dist/client/components/headers");
const mockedHeaders = jest.mocked(headers);

function mockSettings(status: number) {
  mockedGetSettings.mockResolvedValue({
    status: status,
    statusText: "Some Status Text",
    settings: MOCK_SETTINGS,
  });
}

function mockHeaders(
  identifiers?: string,
  keywords?: string,
  identifiersType?: string,
  locale?: string,
) {
  identifiers &&
    when(mockReadOnlyHeaders.get(IDENTIFIERS_HEADER)).thenReturn(identifiers);
  keywords &&
    when(mockReadOnlyHeaders.get(KEYWORDS_HEADER)).thenReturn(keywords);
  identifiersType &&
    when(mockReadOnlyHeaders.get(IDENTIFIER_TYPES_HEADER)).thenReturn(
      identifiersType,
    );
  locale && when(mockReadOnlyHeaders.get(LOCALE_HEADER)).thenReturn(locale);
  mockedHeaders.mockReturnValue(instance(mockReadOnlyHeaders));
}

function mockSearchCatalogItems(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedSearchCatalogItems.mockRejectedValue(error);
  } else {
    mockedSearchCatalogItems.mockResolvedValue(MOCK_CATALOG_ITEMS_API_RESPONSE);
  }
}

describe("Test GET catalogItems API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  afterEach(() => {
    reset(mockReadOnlyHeaders);
  });

  test("returns 200 on successful call to Search Catalog API", async () => {
    mockSettings(200);
    mockHeaders(IDENTIFIERS, undefined, ASIN_IDENTIFIERS_TYPE, US_LOCALE);
    mockSearchCatalogItems();
    mockResolveFetchResponse(200, MOCK_CATALOG_ITEMS_API_RESPONSE);

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 200 on successful call to Search Catalog API with empty response", async () => {
    mockSettings(200);
    mockHeaders(IDENTIFIERS, undefined, ASIN_IDENTIFIERS_TYPE, US_LOCALE);
    mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
    mockedSearchCatalogItems.mockResolvedValue(
      MOCK_CATALOG_ITEMS_API_EMPTY_RESPONSE,
    );
    mockResolveFetchResponse(200, MOCK_CATALOG_ITEMS_API_EMPTY_RESPONSE);

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 200 on successful call to Search Catalog API for SKU Identifiers", async () => {
    mockSettings(200);
    mockHeaders(IDENTIFIERS, "", SKU_IDENTIFIERS_TYPE, US_LOCALE);
    mockSearchCatalogItems();
    mockResolveFetchResponse(200, MOCK_CATALOG_ITEMS_API_RESPONSE);

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 200 on successful call to Search Catalog API for keywords", async () => {
    mockSettings(200);
    mockHeaders(undefined, KEYWORDS, SKU_IDENTIFIERS_TYPE, US_LOCALE);
    mockSearchCatalogItems();
    mockResolveFetchResponse(200, MOCK_CATALOG_ITEMS_API_RESPONSE);

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 500 on exception from settings response when calling Search Catalog API", async () => {
    mockSettings(500);
    mockHeaders(IDENTIFIERS, "", ASIN_IDENTIFIERS_TYPE, US_LOCALE);
    mockSearchCatalogItems();
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await GET(), 500, false);
  });

  test("returns 404 on exception from Search Catalog API", async () => {
    mockSettings(200);
    mockHeaders(undefined, undefined, undefined, US_LOCALE);
    mockSearchCatalogItems({
      status: 404,
    });
    mockRejectFetchResponse(404);

    await verifyStatusAndBody(await GET(), 404, true);
  });
});
