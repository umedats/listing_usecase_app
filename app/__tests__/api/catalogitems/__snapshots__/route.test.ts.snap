// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GET catalogItems API returns 200 on successful call to Search Catalog API 1`] = `
{
  "data": [
    {
      "asin": "B00004SPEC",
      "productType": "KITCHEN",
      "title": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
    },
  ],
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#searchcatalogitems",
      "apiName": "SearchCatalogItems",
      "request": {
        "identifiers": [
          "B00004SPEC",
        ],
        "identifiersType": "ASIN",
        "includedData": [
          "productTypes",
          "summaries",
        ],
        "locale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "items": [
          {
            "asin": "B00004SPEC",
            "productTypes": [
              {
                "marketplaceId": "ATVPDKIKX0DER",
                "productType": "KITCHEN",
              },
            ],
            "summaries": [
              {
                "adultProduct": false,
                "autographed": false,
                "brand": "KRUPS",
                "browseClassification": {
                  "classificationId": "289915",
                  "displayName": "Countertop Blenders",
                },
                "itemClassification": "BASE_PRODUCT",
                "itemName": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
                "manufacturer": "KRUPS",
                "marketplaceId": "ATVPDKIKX0DER",
                "memorabilia": false,
                "modelNumber": "243-70",
                "packageQuantity": 1,
                "partNumber": "243-70",
                "style": "Modern",
                "tradeInEligible": false,
                "websiteDisplayGroup": "kitchen_display_on_website",
                "websiteDisplayGroupName": "Kitchen",
              },
            ],
          },
        ],
        "numberOfResults": 1,
      },
    },
  ],
}
`;

exports[`Test GET catalogItems API returns 200 on successful call to Search Catalog API for SKU Identifiers 1`] = `
{
  "data": [
    {
      "asin": "B00004SPEC",
      "productType": "KITCHEN",
      "title": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
    },
  ],
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#searchcatalogitems",
      "apiName": "SearchCatalogItems",
      "request": {
        "identifiers": [
          "B00004SPEC",
        ],
        "identifiersType": "SKU",
        "includedData": [
          "productTypes",
          "summaries",
        ],
        "locale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
      },
      "response": {
        "items": [
          {
            "asin": "B00004SPEC",
            "productTypes": [
              {
                "marketplaceId": "ATVPDKIKX0DER",
                "productType": "KITCHEN",
              },
            ],
            "summaries": [
              {
                "adultProduct": false,
                "autographed": false,
                "brand": "KRUPS",
                "browseClassification": {
                  "classificationId": "289915",
                  "displayName": "Countertop Blenders",
                },
                "itemClassification": "BASE_PRODUCT",
                "itemName": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
                "manufacturer": "KRUPS",
                "marketplaceId": "ATVPDKIKX0DER",
                "memorabilia": false,
                "modelNumber": "243-70",
                "packageQuantity": 1,
                "partNumber": "243-70",
                "style": "Modern",
                "tradeInEligible": false,
                "websiteDisplayGroup": "kitchen_display_on_website",
                "websiteDisplayGroupName": "Kitchen",
              },
            ],
          },
        ],
        "numberOfResults": 1,
      },
    },
  ],
}
`;

exports[`Test GET catalogItems API returns 200 on successful call to Search Catalog API for keywords 1`] = `
{
  "data": [
    {
      "asin": "B00004SPEC",
      "productType": "KITCHEN",
      "title": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
    },
  ],
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#searchcatalogitems",
      "apiName": "SearchCatalogItems",
      "request": {
        "includedData": [
          "productTypes",
          "summaries",
        ],
        "keywords": [
          "SHIRT",
        ],
        "locale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "items": [
          {
            "asin": "B00004SPEC",
            "productTypes": [
              {
                "marketplaceId": "ATVPDKIKX0DER",
                "productType": "KITCHEN",
              },
            ],
            "summaries": [
              {
                "adultProduct": false,
                "autographed": false,
                "brand": "KRUPS",
                "browseClassification": {
                  "classificationId": "289915",
                  "displayName": "Countertop Blenders",
                },
                "itemClassification": "BASE_PRODUCT",
                "itemName": "Krups 243-70 Power X Plus Combi Blender, DISCONTINUED",
                "manufacturer": "KRUPS",
                "marketplaceId": "ATVPDKIKX0DER",
                "memorabilia": false,
                "modelNumber": "243-70",
                "packageQuantity": 1,
                "partNumber": "243-70",
                "style": "Modern",
                "tradeInEligible": false,
                "websiteDisplayGroup": "kitchen_display_on_website",
                "websiteDisplayGroupName": "Kitchen",
              },
            ],
          },
        ],
        "numberOfResults": 1,
      },
    },
  ],
}
`;

exports[`Test GET catalogItems API returns 200 on successful call to Search Catalog API with empty response 1`] = `
{
  "data": [],
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#searchcatalogitems",
      "apiName": "SearchCatalogItems",
      "request": {
        "identifiers": [
          "B00004SPEC",
        ],
        "identifiersType": "ASIN",
        "includedData": [
          "productTypes",
          "summaries",
        ],
        "locale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "numberOfResults": 0,
      },
    },
  ],
}
`;

exports[`Test GET catalogItems API returns 404 on exception from Search Catalog API 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#searchcatalogitems",
      "apiName": "SearchCatalogItems",
      "request": {
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "status": 404,
      },
    },
  ],
}
`;
